### PR TITLE
IsNullFixer - fix whitespace handling

### PR DESCRIPTION
**Pull request by roukmoute**
Fix https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3057

### CODE REVIEW

File: tests/Fixer/LanguageConstruct/IsNullFixerTest.php [hunk 1]
```diff
@@ -204,6 +205,36 @@ public function provideYodaFixCases()
                 '<?php $result = ((null === $a) <> false); ?>',
                 '<?php $result = (is_null($a) <> false); ?>',
             ],
+
+            // check with logical operator
```
Comment (julienfalque) [inline tests/Fixer/LanguageConstruct/IsNullFixerTest.php]: The next test case does not contain any logical operator, shouldn't the comment be after it?

File: src/Tokenizer/Token.php [hunk 2]
```diff
@@ -474,6 +474,16 @@ public function isWhitespace($whitespaces = " \t\n\r\0\x0B")
     }
 
     /**
+     * Check if token is a logical operator.
+     *
+     * @return bool
+     */
+    public function isLogicalOperator()
```
Comment (julienfalque) [inline src/Tokenizer/Token.php]: Technically, adding this method is a new feature which would require a minor version increment. But this PR is a bugfix that should target a lower branch. Could you please remove it?

**Comment by keradus** [comment]
if 2.2.8, then please change target branch ;)

File: tests/Fixer/LanguageConstruct/IsNullFixerTest.php [hunk 3]
```diff
@@ -76,6 +76,7 @@ public function testNonYodaFix()
             '<?php $b = a(a(a(b() === null) === null) === null) === null;',
             '<?php $b = \is_null(a(\is_null(a(\is_null(a(\is_null(b())))))));'
         );
+        $this->doTest('<?php if ($x === null && $y) echo "foo";', '<?php if (is_null($x) && $y) echo "foo";');
```
Comment (SpacePossum) [inline tests/Fixer/LanguageConstruct/IsNullFixerTest.php]: This is currently the only test that fails without your proposed fixes.

The test passes without your fixes, when you disable this line;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.2.7/src/Fixer/LanguageConstruct/IsNullFixer.php#L106

This makes me think that the `IsNullFixer` should only be changed there by adding a test before the call to `removeTrailingWhitespace`. 

The other test cases are all nice, so :+1: for adding :)

**Comment by roukmoute** [comment]
@juliendufresne @SpacePossum updates done.

I've change branch.
I only remove the use of `removeTrailingWhitespace()`.
I kept all new tests.

**Comment by julienfalque** [comment]
> I only remove the use of `removeTrailingWhitespace()`.

I think you actually forgot to commit that :)

**Comment by roukmoute** [comment]
@juliendufresne I don't understand.
I've commited: `Remove the use of the removeTrailingWhitespace() method`.

Have I miss something ?

**Comment by SpacePossum** [comment]
The target is good now :+1: 
The test file looks good :+1:
Changes in `Token.php` are reverted :+1: 
Sadly, now there is no fix proposal left in the fixer file itself.

**Comment by roukmoute** [comment]
@SpacePossum I have followed your commentary `The test passes without your fixes, when you disable this line;` and this [comment](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3058#discussion_r138987448).
Do you prefer I add a check like before?
Oh perhaps I add a condition before the `removeTrailingWhitespace()` method?

**Comment by roukmoute** [comment]
Oh!
I've just seen that I've completely forgotten to remove `removeTrailingWhitespace()` :-D

I better understand your comments now :)

**Comment by SpacePossum** [comment]
:)
thanks for updating!

>Oh perhaps I add a condition before the removeTrailingWhitespace() method?

I would've to take a better look at before making that call. What I did notice however is that removing all calls to `removeTrailingWhitespace` in the fixer still makes the tests pass. So I wonder, maybe the tests are incomplete or the fixer is doing unneeded things.

**Comment by roukmoute** [comment]
If we try with:
```php
array(
    '<?php if (null === $u ) echo "foo"; ?>',
    '<?php if (is_null($u) ) echo "foo"; ?>',
),
```
we keep a space.
But I don't think it's the job of `IsNullFixer` to remove this space.

Am I right, or I have to improve this P.R.?

**Comment by julienfalque** [comment]
No it's not its job. Some people even like to have spaces inside braces: `if ( is_null($foo) ) { ... }`, the fixer should not remove them, another fixer already takes care of that, if enabled.

**Comment by SteveJobzniak** [comment]
Thanks for this PR. We'll start using this is_null fixer as soon as this fix is merged. :-)

**Comment by keradus** [comment]
Thank you @roukmoute.

**Pull request merged**